contracts/src/RichardsGame/Pool.sol

```diff
@@ -0,0 +1,173 @@
+/**
+ * @notice This is a gambling pool, which lets you gamble with ERC20s and NFTs, similar to csgoempire.co
+ * 1. Users deposit NFTs and ERC20s into the pool and receive a proportional amount of point
+ * 2. Every hour, the pool is spun, a random number between 0 and the total amount of points is generated
+ * 3. The user who has the point wins everything in the pool, you get 30 seconds to claim otherwise rewards stay in the pool
+ */
+import "./interfaces/IPriceOracle.sol";
+import "./interfaces/IRandomNumberGenerator.sol";
+import "../callback/TokenCallbackHandler.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+
+contract Pool is TokenCallbackHandler {
+    mapping(address => tickets) public deposits;
+    struct tickets {
+        uint256 startIndex;
+        uint256 endIndex;
+    }
+    uint256 totalPrevDepositValue;
+    uint256 depositValue;
+    uint256 winningNumber;
+    uint256 withdrawExpiration;
+
+    IPriceOracle public oracle;
+    IRandomNumberGenerator public random;
+    uint256 WITHDRAW_TIME = 30 seconds;
+
+    constructor(IPriceOracle _oracle, IRandomNumberGenerator _random) {
+        oracle = _oracle;
+        random = _random;
+        emit NewOracle(address(_oracle));
+        emit NewRandomNumberGenerator(address(_random));
+    }
+
+    function depositEth() internal returns (uint256 value) {
+        value = oracle.getEthPrice(msg.value);
+        emit DepositEth(msg.sender, msg.value, value);
+    }
+
+    function depositERC20(
+        address token,
+        uint256 amount
+    ) internal returns (uint256 value) {
+        IERC20(token).transferFrom(msg.sender, address(this), amount);
+        value = oracle.getERC20Price(token, amount);
+        emit DepositERC20(msg.sender, token, amount, value);
+    }
+
+    function depositERC721(
+        address token,
+        uint256 tokenId
+    ) internal returns (uint256 value) {
+        IERC721(token).transferFrom(msg.sender, address(this), tokenId);
+        value = oracle.getERC721Price(token, tokenId);
+        emit DepositERC721(msg.sender, token, tokenId, value);
+    }
+
+    function depositAll(
+        address[] calldata erc20,
+        uint256[] calldata amounts,
+        address[] calldata erc721,
+        uint256[] calldata tokendIds
+    ) public payable {
+        require(block.timestamp > withdrawExpiration, "Withdraw not expired");
+        uint256 _depositValue;
+        for (uint256 i = 0; i < erc20.length; i++) {
+            _depositValue += depositERC20(erc20[i], amounts[i]);
+        }
+        for (uint256 i = 0; i < erc721.length; i++) {
+            _depositValue += depositERC721(erc721[i], tokendIds[i]);
+        }
+        _depositValue += depositEth();
+        deposits[msg.sender].startIndex = depositValue;
+        deposits[msg.sender].endIndex = depositValue + _depositValue;
```

Comment:
Just a thought / discussion to open up -

Right now, it looks like the behavior that value / points is purely determined at *time of deposit*, not *time of spin*. In that if I deposit an NFT worth 20 ETH right now, I get 20 points. But at the time of spin, let's say the NFT has risen to 30 ETH. I still have 20 points, since the number of points is determined at the time of spin.

I'm wondering if time of spin could make for some fun mechanics, in that by putting an NFT / ERC-20 token in, you're also betting directionally on that asset's price for future spins.

I think time of deposit makes sense, just

Reply:
too much work I think, but also the price of an NFT shouldn't change that much during that time period between deposit and spin